Implement an automated workflow for working with bootstrap

Ashley-exe opened the pull request:
I've spent the past day putting together a system which will automatically update the ASPX files and web assets within the `AdminSystem` project when you export from bootstrap.

# Get bootstrap
If you haven't already
* Get the github student pack
* find the bootstrap link in the pack
* Connect your github account to bootstrap
* get your pro key
* download and activate bootstrap studio

# Setting up the workflow
Bootstrap has an export hook, which can run a script on every export. We need to set this up independantly.

These settings are unique to each user, so we all have our own `.bsdesign` file - just configure yours.

* Checkout to `meta/bootstrap`
* Open ***your*** `.bsdesign` found in `/BSD/` with bootstrap studio. 
> Export settings are stored in the design, but are unique to each user. We all have our own design.
* Export > dropdown > Export settings
![image](https://user-images.githubusercontent.com/50697488/106364723-14e08700-6329-11eb-8041-6780dd01db8d.png)
* Set the export folder to the BSD location
* under `advanced`, click browse and direct it to `ASPXUpdater\bin\Release\netcoreapp2.1\publish\ASPXUpdater.bat`, ensuring you select the ***BAT***, not the exe - ***and using 2.1, not 3.1*** since y'all have old version of dot net.
* save settings
![image](https://user-images.githubusercontent.com/50697488/106364732-27f35700-6329-11eb-8d1a-1c34e4282010.png)


# The workflow
* In visual studio, open any ASPX in the browser (Right click ASPX > show in browser)
* In bootstrap, create a page ***with the same name as the ASPX*** and make some changes
* click `export`
* when exporting, as command prompt window should open, running my script, updating the ASPX. ***It will not automatically close, unless it encounters a problem***. Read the output.
* Refresh the browser showing the ASPX from visual studio, it should be updated with the bootstrap changes

# Developing
Bootstrap has an in built `preview` server which automatically updates with every change made in the design, so you can use preview to view your page in a browser as you develop. You just need to enable it every time you load the design to start the server, which you can do in the dropdown.
![image](https://user-images.githubusercontent.com/50697488/106365134-64c04d80-632b-11eb-8abf-e26508b0be1f.png)



Once you're happy with your changes, export them and try viewing them in the `AdminSystem` Project.

* Make changes in bootstrap
* export
* `view in browser` on the ASPX from visual studio
OR
* refresh a browser page that's already showing the ASPX  from visual studio after exporting.
![image](https://user-images.githubusercontent.com/50697488/106364812-afd96100-6329-11eb-84da-06280c2d986b.png)
![image](https://user-images.githubusercontent.com/50697488/106364815-b962c900-6329-11eb-9e0d-326e2ba7efb5.png)



# Trouble shooting
If no command window opens when exporting, but a script it assigned in the export settings then bss will likely spit out an `error.log` in `/BSD/`.

* The tool will not create ASPX files, only modify existing ones.
* Name of page in bootstrap must match name of ASPX file
* ASPX files with no matching HTML file will not be updated.
* HTML files with no matching ASPX to compile to will cause an error, and will be ignored.

> _NB_ :  ***Bootstrap design does not automatically save***, not even when exporting. Remember to `ctrl+s` in bootstrap often.


For reference, this is the expected output of a good export:
> This is also saved to the project root as a log file.
```


==================Startup===================


[INFO] Will attempt to work at solution D:\The_Tunnel\Software_Development\C#\CLSG\Skeleton.sln
[INFO] Checking is in project..
[PASS] Is in project!
[INFO] Getting all HTML and ASPX files..
[PASS] Loaded 2 files from D:\The_Tunnel\Software_Development\C#\CLSG\BSD\
[PASS] Loaded 24 files from D:\The_Tunnel\Software_Development\C#\CLSG\AdminSystem\


==================Checking HTML===================


[INFO] HTML 'ThisShouldFail' has no matching ASPX. It will fail, and be ignored.


==================Checking ASPX===================


[INFO] ASPX '1DataEntry' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '1List' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '1Viewer' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '2ConfirmDelete' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '2DataEntry' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '2List' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '2Viewer' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '3ConfirmDelete' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '3DataEntry' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '3List' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '3Viewer' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '4ConfirmDelete' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '4DataEntry' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '4List' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '4Viewer' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '5ConfirmDelete' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '5DataEntry' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '5List' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '5Viewer' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '6ConfirmDelete' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '6DataEntry' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '6List' has no matching HTML in the design. It will not be updated.
[INFO] ASPX '6Viewer' has no matching HTML in the design. It will not be updated.


==================Begin ASPX gen===================


[PASS] Updated ASPX for D:\The_Tunnel\Software_Development\C#\CLSG\BSD\1ConfirmDelete.html
[ERR] Failed to process D:\The_Tunnel\Software_Development\C#\CLSG\BSD\ThisShouldFail.html.
[ERR] Failed to generate ASPX: 'D:\The_Tunnel\Software_Development\C#\CLSG\BSD\ThisShouldFail.html' had no corresponding ASPX file to update. Generate one in Visual Studio, or check file names and paths.
[ERR] STACKTRACE :
   at ASPXUpdater.ASPXUpdater.ProcessFile(String HTMLFile) in D:\The_Tunnel\Software_Development\C#\CLSG\ASPXUpdater\ASPXUpdater.cs:line 130
   at ASPXUpdater.ASPXUpdater.UpdateAllASPX() in D:\The_Tunnel\Software_Development\C#\CLSG\ASPXUpdater\ASPXUpdater.cs:line 107


==================End ASPX gen===================


[PASS] Finished updating ASPX. Updating D:\The_Tunnel\Software_Development\C#\CLSG\AdminSystem\assets\
[INFO] Assets folder already exists in target. Deleting it...
[INFO] Copying over new assets folder...
[PASS] Done!
[PASS] Finished!


==================Idle. Press any key to Halt.===================
```

Commit:
Remove DS_Store

Commit:
Added template bsdesign

Commit:
Many ASPX Updater additions and fixes

Commit:
Updater can be ran automatically on BSS export

adds batch and a shortcut tailored to execute the release build of ASPX updater.

Also force commits the ignored release folder

Commit:
Removed aspx updater shortcut

The shortcut was not relative, but absolute to my machine so it doesn't work for others.

Commit:
Copied design for all devs

Commit:
Fixed missing code, and changed target version

Commit:
Publish for 2.1, and copied batch

Commit:
Changed updater to self contained

anthonyman15 committed:
Implement_Workflow_Successed

anthonyman15 reviewed (approved):
Implement Workflow Complete!

Commit:
The beginnings of HTML to ASP component translation

You cannot code or reference HTML objects from c# in aspx, you need to use ASP objects. Using patterns, i've started on a translator which will transform HTML labels, buttons and text fields to corresponding ASP ones.

Birdie2016 commented:
hello
